url354ce1a2d1_dagaren_gladiator_chess_tgz-pJ8-es_dagaren_gladiator_main_MainJ8/src/es/dagaren/gladiator/representation/BitboardPosition.java

```diff
@@ -0,0 +1,2907 @@
+      //Se devuelve el bitboard de atacadas diagonal A8H1 para la casilla y ocupacion
+      return bbSecondaryDiagonalAttack[squareIndex][occupation];
+   } //Fin de getAtacadasDiagonalA8H1
+   
+   
+   protected void resetMoves()
+   {
+      movesGenerated = false;
+      movesList = null;
+   }
+   
+   
+   /************* MÉTODOS HEREDADOS DE POSICION ABSTRACTA ********************/
+
+   public boolean loadFen(String posicionFen)
+   {
+      resetMoves();
+      
+      boolean result = super.loadFen(posicionFen);
+      
+      updateOcupied();
+      updateAttackedFromFull();
+      
+      return result;
+   }
+   
+   /* (non-Javadoc)
+    * @see es.dagaren.chesi.Posicion#deshacerMovement()
+    */
+   public boolean undoMove(Movement mov)
+   {
+      resetMoves();
+      
+      //Se elimina la posición resultante en el hash de posiciones
+      Integer keyValue = gamePositionsHash.get(zobristKey.getKey());
+      if(keyValue != null)
+      {
+         keyValue = keyValue - 1;
+         if(keyValue == 0)
+            gamePositionsHash.remove(zobristKey.getKey());
+         else
+            gamePositionsHash.put(zobristKey.getKey(), keyValue);
+      }
+      
+      Piece sp = mov.getSourcePiece();
+      Piece dp = mov.getDestinationPiece();
+      Square ss = mov.getSource();
+      Square ds = mov.getDestination();
+      
+      removePiece(ds);
+      if(dp != null)
+         putPiece(dp,ds);
+      putPiece(sp,ss);
+      
+      //Si es captura al paso se pone un peón en la casilla correspondiente 
+      //TODO añadir comprobación de si es casilla al paso en el caso
+      //de que sea un movimiento de peón en diagonal (origen.file != destino.file)
+      //y en la casilla destino no exista pieza
+      if(mov.isInPassant())
+      {
+         if(ds.getRank() == Rank._3)
+         {
+            putPiece(Piece.WHITE_PAWN, enPassantSquareSource[ds.index]);
+         }
+         else if(ds.getRank() == Rank._6)
+         {
+            putPiece(Piece.BLACK_PAWN, enPassantSquareSource[ds.index]);
+         }
+      }
+      
+      if(sp == Piece.WHITE_KING && ss == Square.e1)
+      {
+         if(ds == Square.g1)
+         {
+            //Enroque corto blanco
+            removePiece(Square.f1);
+            putPiece(Piece.WHITE_ROOK,Square.h1);
+            
+         }
+         else if(ds == Square.c1)
+         {
+            //Enroque largo blanco
+            removePiece(Square.d1);
+            putPiece(Piece.WHITE_ROOK,Square.a1);
+         }
+      }
+      else if(sp == Piece.BLACK_KING && ss == Square.e8)
+      {
+         if(ds == Square.g8)
+         {
+            //Enroque corto negro
+            removePiece(Square.f8);
+            putPiece(Piece.BLACK_ROOK,Square.h8);
+         }
+         else if(ds == Square.c8)
+         {
+            //Enroque largo negro
+            removePiece(Square.d8);
+            putPiece(Piece.BLACK_ROOK,Square.a8);
+         }
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1447:1477, 1544:1573.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.